source/common/logger/ur_sinks.hpp

```diff
@@ -25,11 +26,15 @@ class Sink {
     template <typename... Args>
     void log(logger::Level level, const char *fmt, Args &&...args) {
         if (!skip_prefix) {
-            *ostream << "<" << logger_name << ">";
-            *ostream << "[" << level_to_str(level) << "]: ";
+            get_buffer() << "<" << logger_name << ">";
+            get_buffer() << "[" << level_to_str(level) << "]: ";
         }
         format(fmt, std::forward<Args &&>(args)...);
-        *ostream << "\n";
+        get_buffer() << "\n";
+
+        std::scoped_lock<std::mutex> output_lock(output_mutex);
+        *ostream << get_buffer().str();
```

Comment:
Shouldn't this also be under `if (level >= flush_level)`? Otherwise, what's the point of keeping the buffer as thread local if we always write the entire buffer to the stream?

Reply:
What is achieved with current implementation is much better granularity of writing to the output stream. Now the whole message coming from a single log() call goes to an ostream, not a single character as it was before. This has a great impact on printing messages from multiple threads.

I left this out of the flush_level guard for now, as I couldn't make it work properly in cases where there are messages with level lower than the flush level only (thus, during an application run there is no message firing up such a flush). I find it reasonable to leave it this way, as this is also the current behavior of Logger (flushing generally is maintained by an ostream, additionally we can have a guarantee for a flush for specific level messages).

Reply:
Right, but do you need `thread_local` storage, then? If you clear the buffer after each log() anyway, it could be just a local variable, right? Or is it much slower?